### PR TITLE
updated metric error/na/todo such that we can still propegate scoring labels

### PR DESCRIPTION
**Pull request by dgkf**
- updated the `pkg_metric_error`, `pkg_metric_na` and `pkg_metric_todo` to be subclasses of `pkg_metric_condition` (to easier catch them all as one class for special treatment)
- added util function `firstS3method` (inspired by `getS3method`, but finding returning the first valid method for an object's subclasses)
- added labels for all `metric_score.pkg_metric_*` functions
- added `pkg_metric_eval` as a way of encapsulating evaluation of metrics and allow us to capture errors

Now we should have label attributes for all columns in a scored pkg_ref tibble, and all items in a scored pkg_ref list.

### CODE REVIEW

File: R/assess_downloads.R [hunk 1]
```diff
@@ -19,15 +19,16 @@ attr(assess_downloads_1yr, "label") <- "number of downloads in the past year"
...
+    sum(x$downloads$count)
+  })
 }
 
 
 
 #' Defining an Assessment Scoring Function
 #' Score a package for the number of downloads in the past year
-#'
+#'regularized
```
Comment (dgkf) [inline R/assess_downloads.R:31]: Whoops - looks like I had a stray cursor at some point and typed into this comment. Will revert.

File: R/pkg_metric.R [hunk 2]
```diff
@@ -8,14 +8,61 @@
...
+#'   \code{pkg_metric}
+#' @param env An environment in which \code{expr} is to be evaluated
+#' @inheritParams pkg_metric
+#' 
+#' @return a \code{pkg_metric} object containing the result of \code{expr} 
+#'
+pkg_metric_eval <- function(expr, ..., class = c(), env = parent.frame()) {
+  out <- capture_expr_output(substitute(expr), env = env, quoted = TRUE)
+  out_metric <- as_pkg_metric(out, class = class)
+  if (inherits(out, "error")) out_metric <- as_pkg_metric_error(out_metric)
```
Comment (elimillera) [inline R/pkg_metric.R:60]: Could we strip out the '.recording' attribute for erroring metrics?
```
> attr(assess_$downloads_1yr, ".recording")
$expr
{
    sum(x$downloads$count)
}

$attributes
NULL

$visible
[1] TRUE

$traceback
NULL

$output
character(0)
```

File: R/utils.R [hunk 3]
```diff
@@ -161,19 +193,30 @@ capture_expr_output <- function(expr, split = FALSE, env = parent.frame(),
...
 
   structure(
-    if (!quoted) expr_quote else expr,
-    value = res$value,
-    visible = res$visible,
-    traceback = cnds_err_traceback,
-    output = outputs[nzchar(outputs)],
-    class = c("expr_output", "expression"))
+    res$value,
+    .recording = list(
```
Comment (elimillera) [inline R/utils.R:197]: What does this '.recording' value do?

**Review by elimillera — APPROVED** [review]
Looks great. I was thinking of a different way to do this but this works a lot better now that I've had a look at it.

File: R/utils.R (hunk 3 above)
Comment (dgkf) [reply R/utils.R:197]: .recording is a list of console output during the evaluation of the expression. Each line of stdout (from `cat`, `print`) is its own entry in the list, and each `message`, `warning`, or `error` gets its own entry. The print method prints these out as though we're playing back a recording of the execution in an interactive session. 

For now this is just used to capture errors, but we could use something like this to produce a trace of the steps that are executed to collect all the pkg_metric info if we want to generate a report.

I'm not sold on the complexity here, so if it ends up being problematic we can simplify it.

**Comment by marlycormar** [comment]
I second @elimillera's comment =) 

This PR works towards issue #129.

**Pull request merged**